# Day 1 과제

## 실습 1
## 자동차 연비 Data set에서 기술 통계치 구하기 (p25) -data.set : mycars.csv
    * 시내에서 연비(mpg) 통계치 : 구동 방식별 데이터 수, 평균, 중앙값, 표준편차, min,max,Q1,Q3
    * 고속도로에서 연비(highway_mileage) 통계치 : 구동 방식별 데이터 수, 평균, 중앙값, 표준편차, min,max,Q1,Q3

### 필수 라이브러리 추가 및 폰트 설정

In [1]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import scipy.stats

%matplotlib inline
matplotlib.rc('font', family='Malgun Gothic')
plt.rc('font', family = 'Malgun Gothic')

### 데이터 로딩

In [2]:
ds_mycars = pd.read_csv("/home/piai/POSCO_AI_AC/statistic/mycars.csv", engine = "python")
ds_mycars.head()

manufacturer model  displacement  year  cylinder automatic driving  mpg  \
0         audi    a4           1.8  1999         4      auto       f   18   
1         audi    a4           1.8  1999         4    manual       f   21   
2         audi    a4           2.0  2008         4    manual       f   20   
3         audi    a4           2.0  2008         4      auto       f   21   
4         audi    a4           2.8  1999         6      auto       f   16   

   highway_mileage fuel    class  
0               29    p  compact  
1               29    p  compact  
2               31    p  compact  
3               30    p  compact  
4               26    p  compact

### mpg(시내에서의 연비) 통계치 

In [3]:
#관심있는 내용이 구동방식(driving)별 mpg데이터 이므로 두 항목만 mycars.csv데이터에서 추출하여 df로 명명하였다.
df = ds_mycars[['driving','mpg']]
df.groupby('driving').describe()

mpg                                                   
         count       mean       std   min   25%   50%   75%   max
driving                                                          
4        103.0  14.330097  2.874459   9.0  13.0  14.0  16.0  21.0
f        106.0  19.971698  3.626510  11.0  18.0  19.0  21.0  35.0
r         25.0  14.080000  2.215852  11.0  12.0  15.0  15.0  18.0

분석 : driving 방식은 총 3가지 이며 4륜, 전륜, 후륜 구동이 있다.  
데이터에서 4륜은 4, 전륜은 f, 후륜은 r로 표기된다.

위 표에서 count : 데이터 수, mean : 평균, std : 표준편차, min : 최소값, 25% : Q1, 50% : 중앙값, 75% : Q3, max : 최대값을 의미한다.

평균적으로 전륜구동이 평균, 최대, 최소, Q1, 중앙, Q3 값이 다른 방식에 비해 높게 나타난 것을 볼 수 있다. 따라서 전륜 구동이 다른 구동방식에 비해 시내 주행 연비가 좋다는 것을 의미한다. 반면에 4륜구동과 후륜 구동의 경우 평균과 최대 최소값이 서로 유의미한 차이를 보이지 않으며 경우에 따라 우위가 뒤바뀌므로 두 방식간에 유의미한 시내에서의 연비차는 없다고 볼 수 있다.

### highway_mileage(고속도로에서 연비) 통계치

In [4]:
df = ds_mycars[['driving','highway_mileage']]
df.groupby('driving').describe()

highway_mileage                                                   
                  count       mean       std   min   25%   50%   75%   max
driving                                                                   
4                 103.0  19.174757  4.078704  12.0  17.0  18.0  22.0  28.0
f                 106.0  28.160377  4.206881  17.0  26.0  28.0  29.0  44.0
r                  25.0  21.000000  3.662877  15.0  17.0  21.0  24.0  26.0

평균값을 비교했을때 고속도로에서의 연비의 경우 전륜구동이 다른 두 방식에 비해 월등하게 높은 것을 알 수 있다. 또한 최소,Q1,중앙값,Q3,최대값 모두에서 전륜구동이 다른 두 방식의 연비에 비해 높다는 것을 알 수 있으므로 확실하게 전륜 구동이 다른 두 방식보다 고속도로 주행 연비가 좋다는 것을 의미한다. 4륜 구동의 경우 후륜 구동과 비교하였을때 최대 값 수치를 제외한 모든 값에서 후륜구동의 연비보다 안좋다는 것을 알 수 있으나 그 차이가 크지 않음을 알 수 있다. 따라서 3가지 방식을 비교하였을 때 전륜 구동의 고속도로 주행 연비가 월등히 높고 4륜 구동 방식이 후륜 구동 방식보다 미세하게 연비가 낮다는 것을 의미한다.

## 실습2

### 자동차 연비 Data set에서 model, driving 데이터에 대해   
- counts,percents, cumulative counts, cumulative percents 구하기 (p28) -data.set : mycars.csv

### model에 대해

In [5]:
ds_mycars = pd.read_csv("/home/piai/POSCO_AI_AC/statistic/mycars.csv",engine="python")
ds_mycars.head()

df_model = ds_mycars[['model']]
df_model.head()

model
0    a4
1    a4
2    a4
3    a4
4    a4

In [6]:
count = df_model['model'].value_counts().sort_index()
count[:5]

4runner 4wd    6
a4             7
a4 quattro     8
a6 quattro     3
altima         6
Name: model, dtype: int64

In [7]:
cumcnt = np.cumsum(count)
cumcnt[:5]

4runner 4wd     6
a4             13
a4 quattro     21
a6 quattro     24
altima         30
Name: model, dtype: int64

In [8]:
percent = count/sum(count)*100
percent[:5]

4runner 4wd    2.564103
a4             2.991453
a4 quattro     3.418803
a6 quattro     1.282051
altima         2.564103
Name: model, dtype: float64

In [9]:
cumpct = np.cumsum(percent)
cumpct[:5]

4runner 4wd     2.564103
a4              5.555556
a4 quattro      8.974359
a6 quattro     10.256410
altima         12.820513
Name: model, dtype: float64

In [10]:
count_data = pd.DataFrame({'Count':count, 
                           'Percent':percent,
                           'Cumulative counts':cumcnt,
                           'Cumulative percents':cumpct})
count_data.columns.name = "model"
count_data

model                   Count   Percent  Cumulative counts  \
4runner 4wd                 6  2.564103                  6   
a4                          7  2.991453                 13   
a4 quattro                  8  3.418803                 21   
a6 quattro                  3  1.282051                 24   
altima                      6  2.564103                 30   
c1500 suburban 2wd          5  2.136752                 35   
camry                       7  2.991453                 42   
camry solara                7  2.991453                 49   
caravan 2wd                11  4.700855                 60   
civic                       9  3.846154                 69   
corolla                     5  2.136752                 74   
corvette                    5  2.136752                 79   
dakota pickup 4wd           9  3.846154                 88   
durango 4wd                 7  2.991453                 95   
expedition 2wd              3  1.282051                 98   
explorer 4wd                6  2.564103                104   
f150 pickup 4wd             7  2.991453                111   
forester awd                6  2.564103                117   
grand cherokee 4wd          8  3.418803                125   
grand prix                  5  2.136752                130   
gti                         5  2.136752                135   
impreza awd                 8  3.418803                143   
jetta                       9  3.846154                152   
k1500 tahoe 4wd             4  1.709402                156   
land cruiser wagon 4wd      2  0.854701                158   
malibu                      5  2.136752                163   
maxima                      3  1.282051                166   
mountaineer 4wd             4  1.709402                170   
mustang                     9  3.846154                179   
navigator 2wd               3  1.282051                182   
new beetle                  6  2.564103                188   
passat                      7  2.991453                195   
pathfinder 4wd              4  1.709402                199   
ram 1500 pickup 4wd        10  4.273504                209   
range rover                 4  1.709402                213   
sonata                      7  2.991453                220   
tiburon                     7  2.991453                227   
toyota tacoma 4wd           7  2.991453                234   

model                   Cumulative percents  
4runner 4wd                        2.564103  
a4                                 5.555556  
a4 quattro                         8.974359  
a6 quattro                        10.256410  
altima                            12.820513  
c1500 suburban 2wd                14.957265  
camry                             17.948718  
camry solara                      20.940171  
caravan 2wd                       25.641026  
civic                             29.487179  
corolla                           31.623932  
corvette                          33.760684  
dakota pickup 4wd                 37.606838  
durango 4wd                       40.598291  
expedition 2wd                    41.880342  
explorer 4wd                      44.444444  
f150 pickup 4wd                   47.435897  
forester awd                      50.000000  
grand cherokee 4wd                53.418803  
grand prix                        55.555556  
gti                               57.692308  
impreza awd                       61.111111  
jetta                             64.957265  
k1500 tahoe 4wd                   66.666667  
land cruiser wagon 4wd            67.521368  
malibu                            69.658120  
maxima                            70.940171  
mountaineer 4wd                   72.649573  
mustang                           76.495726  
navigator 2wd                     77.777778  
new beetle                        80.341880  
passat                            83.333333  
pathfinder 4wd                    85.042735  
ram 1500 pickup 4

mycars.csv 데이터는 총 234대의 차량의 정보를 담고 있다. 모델을 기준으로 하여 counts, percents, cumulative counts, cumulative percents를 구하였을 때 가장 많은 모델은 caravan 2wd 이고 가장 적은 모델은 land cruiser wagon 4wd이다. 이 분류를 통해 이 데이터에 차량 모델 별 수량과 전체 중 몇 퍼센트를 차지하는지 알 수 있었다. 

### driving에 대해

In [11]:
ds_mycars = pd.read_csv("/home/piai/POSCO_AI_AC/statistic/mycars.csv",engine="python")
ds_mycars.head()

df_driving = ds_mycars[['driving']]
df_driving.head()

driving
0       f
1       f
2       f
3       f
4       f

In [12]:
count = df_driving['driving'].value_counts().sort_index()
count

4    103
f    106
r     25
Name: driving, dtype: int64

In [13]:
cumcnt = np.cumsum(count)
cumcnt

4    103
f    209
r    234
Name: driving, dtype: int64

In [14]:
percent = count/sum(count)*100
percent

4    44.017094
f    45.299145
r    10.683761
Name: driving, dtype: float64

In [15]:
cumpct = np.cumsum(percent)
cumpct[:5]

4     44.017094
f     89.316239
r    100.000000
Name: driving, dtype: float64

In [16]:
count_data = pd.DataFrame({'Count':count, 
                           'Percent':percent,
                           'Cumulative counts':cumcnt,
                           'Cumulative percents':cumpct})
count_data.columns.name = "driving"
count_data

driving  Count    Percent  Cumulative counts  Cumulative percents
4          103  44.017094                103            44.017094
f          106  45.299145                209            89.316239
r           25  10.683761                234           100.000000

mycars.csv 파일의 자료를 구동 방식의 분포에 따라 정리하였다. 구동 방식은 총 3가지(4륜, 전륜, 후륜구동)가 있으며 전륜 구동 차량이 가장 많으나 4륜 구동 차량 또한 비슷하게 많았으며 후륜 구동 차량이 가장 적었다. 이 분류를 통해 이 데이터에 구동방식 별 차량 수와 전체에서 몇퍼센트를 차지하는지 알 수 있었다.